bionic/code_hasher.py

```diff
@@ -0,0 +1,201 @@
+"""
+This module contains the CodeHasher class, which hashes code objects
+into values that uniquely identify those objects.
+
+Bionic uses the CodeHasher class to detect any changes to entity
+functions. When an entity function changes, the entity value may also
+change, which requires Bionic to invalidate the cache and compute the
+entity again.
+
+This class is only used when versioning mode is set to "auto" or
+"assisted". Since the logic doesn't cover every type of code change,
+like changes to references or classes, these versioning modes are still
+experimental. When we can detect all kinds of code changes, we will
+make "auto" mode the default behavior.
+"""
+
+from enum import Enum
+import hashlib
+import inspect
+import warnings
+
+from .utils.misc import oneline
+
+
+PREFIX_SEPARATOR = b"$"
+
+
+class CodeHasher:
+    """
+    This class hashes code objects into values that uniquely identify
+    those objects. If two code objects have different behavior, they
+    should ideally have different hashes. However, for now, we only
+    examine the function bytecode and associated constants. We will add
+    more functionality like checking function references and any
+    changes to class objects.
+
+    Since it doesn't examines the references and class objects, it
```

Comment:
s/examines/examine/

Also, it's confusing to use "it" here without making it clear what the referent is.